File: LR3/Realization_GO/singly_linked_list.go
package main

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"os"
	"strconv"
)

// Node - узел односвязного списка
type Node struct {
	Key  int
	Next *Node
}

// SinglyLinkedList - односвязный список
type SinglyLinkedList struct {
	Head *Node
}

// NewSinglyLinkedList - конструктор
func NewSinglyLinkedList() *SinglyLinkedList {
	return &SinglyLinkedList{
		Head: nil,
	}
}

// createNode - создание нового узла
func (s *SinglyLinkedList) createNode(value int) *Node {
	return &Node{
		Key:  value,
		Next: nil,
	}
}

// AddToHead - добавление в начало
func (s *SinglyLinkedList) AddToHead(value int) {
	newNode := s.createNode(value)
	newNode.Next = s.Head
	s.Head = newNode
	fmt.Printf("Добавлен в начало: %d\n", value)
}

// AddToTail - добавление в конец
func (s *SinglyLinkedList) AddToTail(value int) {
	newNode := s.createNode(value)

	if s.Head == nil {
		s.Head = newNode
		fmt.Printf("Добавлен в конец: %d\n", value)
		return
	}

	current := s.Head
	for current.Next != nil {
		current = current.Next
	}
	current.Next = newNode
	fmt.Printf("Добавлен в конец: %d\n", value)
}

// AddAfter - добавление после значения
func (s *SinglyLinkedList) AddAfter(targetValue, newValue int) {
	if s.Head == nil {
		fmt.Println("Список пуст")
		return
	}

	current := s.Head
	for current != nil && current.Key != targetValue {
		current = current.Next
	}

	if current == nil {
		fmt.Printf("%d не найден!\n", targetValue)
		return
	}

	newNode := s.createNode(newValue)
	newNode.Next = current.Next
	current.Next = newNode
	fmt.Printf("Добавлен после %d: %d\n", targetValue, newValue)
}

// AddBefore - добавление перед значением
func (s *SinglyLinkedList) AddBefore(targetValue, newValue int) {
	if s.Head == nil {
		fmt.Println("Список пуст")
		return
	}

	if s.Head.Key == targetValue {
		s.AddToHead(newValue)
		return
	}

	current := s.Head
	for current.Next != nil && current.Next.Key != targetValue {
		current = current.Next
	}

	if current.Next == nil {
		fmt.Printf("%d не найден!\n", targetValue)
		return
	}

	newNode := s.createNode(newValue)
	newNode.Next = current.Next
	current.Next = newNode
	fmt.Printf("Добавлен перед %d: %d\n", targetValue, newValue)
}

// DeleteBefore - удаление перед значением
func (s *SinglyLinkedList) DeleteBefore(targetValue int) {
	if s.Head == nil || s.Head.Next == nil {
		fmt.Println("Недостаточно элементов для удаления!")
		return
	}

	if s.Head.Key == targetValue {
		fmt.Println("Невозможно удалить элемент перед первым элементом!")
		return
	}

	if s.Head.Next.Key == targetValue {
		temp := s.Head
		s.Head = s.Head.Next
		fmt.Printf("Удален элемент: %d перед элементом %d\n", temp.Key, targetValue)
		return
	}

	current := s.Head
	for current.Next.Next != nil && current.Next.Next.Key != targetValue {
		current = current.Next
	}

	if current.Next.Next == nil {
		fmt.Printf("Элемент %d не найден!\n", targetValue)
		return
	}

	nodeToDelete := current.Next
	current.Next = current.Next.Next
	fmt.Printf("Удален элемент: %d перед элементом %d\n", nodeToDelete.Key, targetValue)
}

// DeleteAfter - удаление после значения
func (s *SinglyLinkedList) DeleteAfter(targetValue int) {
	if s.Head == nil {
		fmt.Println("Список пуст! Невозможно удалить элемент.")
		return
	}

	current := s.Head
	for current != nil && current.Key != targetValue {
		current = current.Next
	}

	if current == nil {
		fmt.Printf("Элемент %d не найден в списке!\n", targetValue)
		return
	}

	if current.Next == nil {
		fmt.Printf("После элемента %d нет элементов для удаления!\n", targetValue)
		return
	}

	nodeToDelete := current.Next
	current.Next = current.Next.Next
	fmt.Printf("Удален элемент: %d после элемента %d\n", nodeToDelete.Key, targetValue)
}

// Print - печать списка
func (s *SinglyLinkedList) Print() {
	if s.Head == nil {
		fmt.Println("Список пуст")
		return
	}
	current := s.Head
	for current != nil {
		fmt.Printf("%d -> ", current.Key)
		current = current.Next
	}
	fmt.Println("nullptr")
}

// PrintReverse - печать в обратном порядке
func (s *SinglyLinkedList) PrintReverse() {
	if s.Head == nil {
		fmt.Println("Список пуст")
		return
	}

	stack := []int{}
	current := s.Head

	for current != nil {
		stack = append(stack, current.Key)
		current = current.Next
	}

	fmt.Print("Список в обратном порядке: ")
	for i := len(stack) - 1; i >= 0; i-- {
		fmt.Printf("%d -> ", stack[i])
	}
	fmt.Println("nullptr")
}

// DeleteFromHead - удаление из начала
func (s *SinglyLinkedList) DeleteFromHead() {
	if s.Head == nil {
		fmt.Println("Список пуст!")
		return
	}
	temp := s.Head
	s.Head = s.Head.Next
	fmt.Printf("Удален из начала: %d\n", temp.Key)
}

// DeleteFromTail - удаление из конца
func (s *SinglyLinkedList) DeleteFromTail() {
	if s.Head == nil {
		fmt.Println("Список пуст!")
		return
	}

	if s.Head.Next == nil {
		fmt.Printf("Удален из конца: %d\n", s.Head.Key)
		s.Head = nil
		return
	}

	current := s.Head
	for current.Next.Next != nil {
		current = current.Next
	}

	temp := current.Next
	current.Next = nil
	fmt.Printf("Удален из конца: %d\n", temp.Key)
}

// DeleteByValue - удаление по значению
func (s *SinglyLinkedList) DeleteByValue(value int) {
	if s.Head == nil {
		fmt.Println("Список пуст!")
		return
	}

	// Удаление головы
	if s.Head.Key == value {
		temp := s.Head
		s.Head = s.Head.Next
		fmt.Printf("Удален элемент: %d\n", temp.Key)
		return
	}

	// Удаление не головы
	current := s.Head
	var prev *Node

	for current != nil && current.Key != value {
		prev = current
		current = current.Next
	}

	if current == nil {
		fmt.Printf("Элемент %d не найден!\n", value)
		return
	}

	if prev != nil {
		prev.Next = current.Next
	}

	fmt.Printf("Удален элемент: %d\n", current.Key)
}

// FindByValue - поиск по значению
func (s *SinglyLinkedList) FindByValue(value int) bool {
	current := s.Head
	for current != nil {
		if current.Key == value {
			return true
		}
		current = current.Next
	}
	return false
}

// Clear - очистка списка
func (s *SinglyLinkedList) Clear() {
	s.Head = nil
}

// SaveToFile - сохранение в текстовый файл
func (s *SinglyLinkedList) SaveToFile(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Printf("Ошибка создания файла: %s\n", filename)
		return err
	}
	defer file.Close()

	current := s.Head
	for current != nil {
		_, err := fmt.Fprintf(file, "%d ", current.Key)
		if err != nil {
			return err
		}
		current = current.Next
	}

	fmt.Printf("Односвязный список сохранён в файл: %s\n", filename)
	return nil
}

// LoadFromFile - загрузка из текстового файла
func (s *SinglyLinkedList) LoadFromFile(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия файла: %s\n", filename)
		return err
	}
	defer file.Close()

	s.Clear()

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanWords)

	for scanner.Scan() {
		value, err := strconv.Atoi(scanner.Text())
		if err == nil {
			s.AddToTail(value)
		}
	}

	fmt.Printf("Односвязный список загружен из файла: %s\n", filename)
	return scanner.Err()
}

// SaveToBinaryFile - сохранение в бинарный файл
func (s *SinglyLinkedList) SaveToBinaryFile(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия бинарного файла для записи: %s\n", filename)
		return err
	}
	defer file.Close()

	count := 0
	current := s.Head
	for current != nil {
		count++
		current = current.Next
	}

	err = binary.Write(file, binary.LittleEndian, int32(count))
	if err != nil {
		return err
	}

	current = s.Head
	for current != nil {
		err = binary.Write(file, binary.LittleEndian, int32(current.Key))
		if err != nil {
			return err
		}
		current = current.Next
	}

	fmt.Printf("Односвязный список сохранён в бинарный файл: %s\n", filename)
	return nil
}

// LoadFromBinaryFile - загрузка из бинарного файла
func (s *SinglyLinkedList) LoadFromBinaryFile(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия бинарного файла для чтения: %s\n", filename)
		return err
	}
	defer file.Close()

	s.Clear()

	var count int32
	err = binary.Read(file, binary.LittleEndian, &count)
	if err != nil {
		return err
	}

	for i := 0; i < int(count); i++ {
		var value int32
		err = binary.Read(file, binary.LittleEndian, &value)
		if err != nil {
			return err
		}
		s.AddToTail(int(value))
	}

	fmt.Printf("Односвязный список загружен из бинарного файла: %s\n", filename)
	return nil
}

File: LR3/Realization_C++/SinglyLinkedList.cpp
#include "SinglyLinkedList.h"
#include <iostream>
#include <fstream>
#include <stack>

using namespace std;

Node* SinglyLinkedList::createNode(int value) {
    Node* newNode = new Node;
    newNode->key = value;
    newNode->next = nullptr;
    return newNode;
}

SinglyLinkedList::SinglyLinkedList() : head(nullptr) {}

SinglyLinkedList::~SinglyLinkedList() {
    clear();
}

void SinglyLinkedList::addToHead(int value) {
    Node* newNode = createNode(value);
    newNode->next = head;
    head = newNode;
    cout << "Добавлен в начало: " << value << endl;
}

void SinglyLinkedList::addToTail(int value) {
    Node* newNode = createNode(value);
    
    if(head == nullptr) {
        head = newNode;
        cout << "Добавлен в конец: " << value << endl;
        return;
    }

    Node* current = head;
    while (current->next != nullptr) {
        current = current->next;
    }
    current->next = newNode;
    cout << "Добавлен в конец: " << value << endl;
}

void SinglyLinkedList::addAfter(int targetValue, int newValue) {
    if (head == nullptr) {
        cout << "Список пуст" << endl;
        return;
    }

    Node* current = head;
    while (current != nullptr && current->key != targetValue) {
        current = current->next;
    }

    if (current == nullptr) {
        cout << targetValue << " не найден!" << endl;
        return;
    }

    Node* newNode = createNode(newValue);
    newNode->next = current->next;
    current->next = newNode;
    cout << "Добавлен после " << targetValue << ": " << newValue << endl;
}

void SinglyLinkedList::addBefore(int targetValue, int newValue) {
    if (head == nullptr) {
        cout << "Список пуст" << endl;
        return;
    }

    if (head->key == targetValue) {
        addToHead(newValue);
        return;
    }

    Node* current = head;
    while (current->next != nullptr && current->next->key != targetValue) {
        current = current->next;
    }

    if (current->next == nullptr) {
        cout << targetValue << " не найден!" << endl;
        return;
    }

    Node* newNode = createNode(newValue);
    newNode->next = current->next;
    current->next = newNode;
    cout << "Добавлен перед " << targetValue << ": " << newValue << endl;
}

void SinglyLinkedList::deleteBefore(int targetValue) {
    if (head == nullptr || head->next == nullptr) {
        cout << "Недостаточно элементов для удаления!" << endl;
        return;
    }

    if (head->key == targetValue) {
        cout << "Невозможно удалить элемент перед первым элементом!" << endl;
        return;
    }

    if (head->next->key == targetValue) {
        Node* temp = head;
        head = head->next;
        cout << "Удален элемент: " << temp->key << " перед элементом " << targetValue << endl;
        delete temp;
        return;
    }

    Node* current = head;
    while (current->next->next != nullptr && current->next->next->key != targetValue) {
        current = current->next;
    }

    if (current->next->next == nullptr) {
        cout << "Элемент " << targetValue << " не найден!" << endl;
        return;
    }

    Node* nodeToDelete = current->next;
    current->next = current->next->next;
    cout << "Удален элемент: " << nodeToDelete->key << " перед элементом " << targetValue << endl;
    delete nodeToDelete;
}

void SinglyLinkedList::deleteAfter(int targetValue) {
    if (head == nullptr) {
        cout << "Список пуст! Невозможно удалить элемент." << endl;
        return;
    }
    
    Node* current = head;
    
    while (current != nullptr && current->key != targetValue) {
        current = current->next;
    }
    
    if (current == nullptr) {
        cout << "Элемент " << targetValue << " не найден в списке!" << endl;
        return;
    }
    
    if (current->next == nullptr) {
        cout << "После элемента " << targetValue << " нет элементов для удаления!" << endl;
        return;
    }
    
    Node* nodeToDelete = current->next;
    current->next = current->next->next;
    
    cout << "Удален элемент: " << nodeToDelete->key << " после элемента " << targetValue << endl;
    delete nodeToDelete;
}

void SinglyLinkedList::print() {
    if (head == nullptr) {
        cout << "Список пуст" << endl;
        return;
    }
    Node* current = head;
    while (current != nullptr) {
        cout << current->key << " -> ";
        current = current->next;
    }
    cout << "nullptr" << endl;
}

void SinglyLinkedList::printReverse() {
    if (head == nullptr) {
        cout << "Список пуст" << endl;
        return;
    }
    
    stack<int> st;
    Node* current = head;
    
    while (current != nullptr) {
        st.push(current->key);
        current = current->next;
    }
    
    cout << "Список в обратном порядке: ";
    while (!st.empty()) {
        cout << st.top() << " -> ";
        st.pop();
    }
    cout << "nullptr" << endl;
}

void SinglyLinkedList::deleteFromHead() {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }
    Node* temp = head;
    head = head->next;
    cout << "Удален из начала: " << temp->key << endl;
    delete temp;
}

void SinglyLinkedList::deleteFromTail() {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }
    
    if (head->next == nullptr) {
        cout << "Удален из конца: " << head->key << endl;
        delete head;
        head = nullptr;
        return;
    }
    
    Node* current = head;
    while (current->next->next != nullptr) {
        current = current->next;
    }
    
    Node* temp = current->next;
    current->next = nullptr;
    cout << "Удален из конца: " << temp->key << endl;
    delete temp;
}

void SinglyLinkedList::deleteByValue(int value) {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }

    if (head->key == value) {
        Node* temp = head;
        head = head->next;
        cout << "Удален элемент: " << temp->key << endl;
        delete temp;
        return;
    }
    
    Node* current = head;
    Node* prev = nullptr;
    
    while (current != nullptr && current->key != value) {
        prev = current;
        current = current->next;
    }
    
    if (current == nullptr) {
        cout << "Элемент " << value << " не найден!" << endl;
        return;
    }
    
    if (prev != nullptr) {
        prev->next = current->next;
    }
    
    cout << "Удален элемент: " << current->key << endl;
    delete current;
}

bool SinglyLinkedList::findByValue(int value) {
    Node* current = head;
    while (current != nullptr) {
        if (current->key == value) {
            return true;
        }
        current = current->next;
    }
    return false;
}

void SinglyLinkedList::clear() {
    while (head != nullptr){
        Node* temp = head;
        head = head->next;
        delete temp;
    }
}

void SinglyLinkedList::saveToFile(const string& filename) {
    ofstream file(filename);
    if(file.is_open()) {
        Node* current = head;
        while(current != nullptr) {
            file << current->key << " ";
            current = current->next;
        }
        file.close();
        cout << "Односвязный список сохранён в файл: " << filename << endl;
    }
}

void SinglyLinkedList::loadFromFile(const string& filename) {
    ifstream file(filename);
    if(file.is_open()) {
        clear();
        int value;
        while(file >> value) {
            addToTail(value);
        }
        file.close();
        cout << "Односвязный список загружен из файла: " << filename << endl;
    }
}

void SinglyLinkedList::saveToBinaryFile(const string& filename) {
    ofstream file(filename, ios::binary);
    if(file.is_open()) {
        Node* current = head;
        int count = 0;
        
        while(current != nullptr) {
            count++;
            current = current->next;
        }
        
        file.write(reinterpret_cast<const char*>(&count), sizeof(int));
        
        current = head;
        while(current != nullptr) {
            file.write(reinterpret_cast<const char*>(&current->key), sizeof(int));
            current = current->next;
        }
        
        file.close();
        cout << "Односвязный список сохранён в бинарный файл: " << filename << endl;
    } else {
        cout << "Ошибка открытия бинарного файла для записи: " << filename << endl;
    }
}

void SinglyLinkedList::loadFromBinaryFile(const string& filename) {
    ifstream file(filename, ios::binary);
    if(file.is_open()) {
        clear();
        
        int count = 0;
        file.read(reinterpret_cast<char*>(&count), sizeof(int));
        
        for(int i = 0; i < count; i++) {
            int value;
            file.read(reinterpret_cast<char*>(&value), sizeof(int));
            addToTail(value);
        }
        
        file.close();
        cout << "Односвязный список загружен из бинарного файла: " << filename << endl;
    } else {
        cout << "Ошибка открытия бинарного файла для чтения: " << filename << endl;
    }
}
